File: plugins/storage/inmemory-py/harness_test.py
"""Conformance/golden-data harness for the storage/v1 axis — Python side.

The ADR-003 cross-run: loads the SAME golden vectors the Go reference loads
(contracts/conformance/storage-v1.json) and drives THIS independent implementation's
StorageService.VendCredentials over real gRPC. The harness sets the caller's tenant
in the rat-callmeta-bin metadata header (ADR-007); the server reads it to scope the
vended credentials; the harness decodes the (conformance) scope receipt and asserts
the C7 obligation (scope.tenant == caller tenant, + prefix + mode + short TTL).

Runs standalone (`python harness_test.py`) or under pytest (`test_*`).
"""

import json
import os
import time
from concurrent import futures

import grpc

from rat.common.v1 import context_pb2
from rat.storage.v1 import storage_pb2, storage_pb2_grpc

from server import StorageServicer

VECTOR_PATH = os.path.join(
    os.path.dirname(__file__), "..", "..", "..", "contracts", "conformance", "storage-v1.json"
)

_CODE = {
    "INVALID_ARGUMENT": grpc.StatusCode.INVALID_ARGUMENT,
    "PERMISSION_DENIED": grpc.StatusCode.PERMISSION_DENIED,
    "NOT_FOUND": grpc.StatusCode.NOT_FOUND,
}

_MODE = {
    "READ": storage_pb2.AccessMode.ACCESS_MODE_READ,
    "WRITE": storage_pb2.AccessMode.ACCESS_MODE_WRITE,
    "READ_WRITE": storage_pb2.AccessMode.ACCESS_MODE_READ_WRITE,
    "UNSPECIFIED": storage_pb2.AccessMode.ACCESS_MODE_UNSPECIFIED,
}


def _now_ms() -> int:
    return int(time.time() * 1000)


def load_vectors():
    with open(VECTOR_PATH, "r", encoding="utf-8") as f:
        v = json.load(f)
    assert v["axis"] == "storage/v1", f'vectors axis = {v["axis"]!r}, want storage/v1'
    return v


class Rig:
    def __init__(self, tenant: str) -> None:
        self.tenant = tenant
        self.server = grpc.server(futures.ThreadPoolExecutor(max_workers=4))
        storage_pb2_grpc.add_StorageServiceServicer_to_server(StorageServicer(), self.server)
        port = self.server.add_insecure_port("127.0.0.1:0")
        self.server.start()
        self.channel = grpc.insecure_channel(f"127.0.0.1:{port}")
        self.stub = storage_pb2_grpc.StorageServiceStub(self.channel)

    def close(self) -> None:
        self.channel.close()
        self.server.stop(None)

    def _callmeta(self):
        rc = context_pb2.RequestContext(
            trace=context_pb2.TraceContext(
                traceparent="00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
                correlation_id="corr-golden",
            ),
            identity=context_pb2.Identity(tenant=self.tenant),
        )
        return [("rat-callmeta-bin", rc.SerializeToString())]

    def vend(self, prefix, mode):
        return self.stub.VendCredentials(
            storage_pb2.VendCredentialsRequest(prefix=prefix, mode=_MODE[mode]),
            metadata=self._callmeta(),
        )


def run_lifecycle(rig: Rig, v) -> None:
    ttl_ms = v["credentials_ttl_seconds"] * 1000
    slack_ms = 2000
    for s in v["lifecycle"]:
        expect = s["expect"]
        before = _now_ms()
        resp = rig.vend(s["prefix"], s["mode"])
        after = _now_ms()
        if expect.get("credentials_present"):
            assert resp.credentials, "credentials empty, want present"
        got = json.loads(resp.credentials.decode("utf-8"))
        sc = expect.get("scope")
        if sc:
            assert got["tenant"] == sc["tenant"], f'scope.tenant = {got["tenant"]!r}, want {sc["tenant"]!r}'
            assert got["prefix"] == sc["prefix"], f'scope.prefix = {got["prefix"]!r}, want {sc["prefix"]!r}'
            assert got["mode"] == sc["mode"], f'scope.mode = {got["mode"]!r}, want {sc["mode"]!r}'
        exp = resp.expires_unix_ms
        assert before + ttl_ms - slack_ms <= exp <= after + ttl_ms + slack_ms, (
            f"expires_unix_ms = {exp}, want ~ now + {ttl_ms}ms"
        )
        assert exp == got["expires_unix_ms"], "response expires != receipt expires"


def run_errors(rig: Rig, v) -> None:
    for s in v["errors"]:
        want = _CODE[s["expect"]["code"]]
        try:
            rig.vend(s["prefix"], s["mode"])
        except grpc.RpcError as e:
            assert e.code() == want, f'{s["step"]}: status = {e.code()}, want {want}'
        else:
            raise AssertionError(f'{s["step"]}: want error {s["expect"]["code"]}, got success')


def test_golden_vectors():
    v = load_vectors()
    rig = Rig(v["tenant"])
    try:
        run_lifecycle(rig, v)
    finally:
        rig.close()


def test_error_vectors():
    v = load_vectors()
    rig = Rig(v["tenant"])
    try:
        run_errors(rig, v)
    finally:
        rig.close()


def test_tenant_from_metadata_not_request():
    """C7 structural check: the vended tenant tracks the metadata envelope; there is
    no request field that could override it."""
    rig = Rig("globex")
    try:
        resp = rig.vend("s3://bucket/x", "READ")
        got = json.loads(resp.credentials.decode("utf-8"))
        assert got["tenant"] == "globex", f'scope.tenant = {got["tenant"]!r}, want globex (the metadata tenant)'
    finally:
        rig.close()


if __name__ == "__main__":
    v = load_vectors()
    rig = Rig(v["tenant"])
    try:
        run_lifecycle(rig, v)
        run_errors(rig, v)
    finally:
        rig.close()
    test_tenant_from_metadata_not_request()
    print("PASS — rat-storage-inmemory-py conformed to storage/v1 golden vectors")

File: plugins/storage/inmemory-go/harness_test.go
// harness_test.go — conformance/golden-data harness for the storage/v1 axis.
//
// Loads contracts/conformance/storage-v1.json (the SAME file the Python reference
// loads) and drives StorageService.VendCredentials through the stub core-mediated
// gateway. The harness sets the caller's tenant in the rat-callmeta-bin envelope;
// the gateway re-stamps it downstream; the storage server reads it to scope the
// vended credentials. The harness then decodes the (conformance) scope receipt and
// asserts the C7 obligation: vended scope.tenant == the caller's tenant, plus the
// requested prefix + mode, plus a short TTL.
package main

import (
	"context"
	"encoding/json"
	"net"
	"os"
	"testing"
	"time"

	commonv1 "github.com/squat-collective/rat-v3/contracts/sdks/go/rat/common/v1"
	corev1 "github.com/squat-collective/rat-v3/contracts/sdks/go/rat/core/v1"
	storagev1 "github.com/squat-collective/rat-v3/contracts/sdks/go/rat/storage/v1"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
	"google.golang.org/grpc/test/bufconn"
	"google.golang.org/protobuf/proto"
)

// ---- golden-vector model (mirrors contracts/conformance/storage-v1.json) ----

type scope struct {
	Tenant string `json:"tenant"`
	Prefix string `json:"prefix"`
	Mode   string `json:"mode"`
}

type expectation struct {
	CredentialsPresent bool   `json:"credentials_present"`
	Scope              *scope `json:"scope"`
	Code               string `json:"code"`
}

type vstep struct {
	Step   string      `json:"step"`
	Op     string      `json:"op"`
	Prefix string      `json:"prefix"`
	Mode   string      `json:"mode"`
	Expect expectation `json:"expect"`
}

type vectors struct {
	Axis            string  `json:"axis"`
	Tenant          string  `json:"tenant"`
	CredentialsTTLs int64   `json:"credentials_ttl_seconds"`
	Lifecycle       []vstep `json:"lifecycle"`
	Errors          []vstep `json:"errors"`
}

const vectorPath = "../../../contracts/conformance/storage-v1.json"

func loadVectors(t *testing.T) vectors {
	t.Helper()
	raw, err := os.ReadFile(vectorPath)
	if err != nil {
		t.Fatalf("read golden vectors %s: %v", vectorPath, err)
	}
	var v vectors
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("parse golden vectors: %v", err)
	}
	if v.Axis != "storage/v1" {
		t.Fatalf("vectors axis = %q, want storage/v1", v.Axis)
	}
	return v
}

func modeEnum(s string) storagev1.AccessMode {
	switch s {
	case "READ":
		return storagev1.AccessMode_ACCESS_MODE_READ
	case "WRITE":
		return storagev1.AccessMode_ACCESS_MODE_WRITE
	case "READ_WRITE":
		return storagev1.AccessMode_ACCESS_MODE_READ_WRITE
	default:
		return storagev1.AccessMode_ACCESS_MODE_UNSPECIFIED
	}
}

// ---- harness wiring: plugin behind the stub core gateway ----

type rig struct {
	gw     corev1.CapabilityInvokeServiceClient
	core   *stubGateway
	tenant string
}

func bufDial(t *testing.T, lis *bufconn.Listener) *grpc.ClientConn {
	t.Helper()
	conn, err := grpc.NewClient(
		"passthrough:///bufnet",
		grpc.WithContextDialer(func(_ context.Context, _ string) (net.Conn, error) { return lis.Dial() }),
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		t.Fatalf("dial: %v", err)
	}
	t.Cleanup(func() { _ = conn.Close() })
	return conn
}

func newRig(t *testing.T, tenant string) *rig {
	t.Helper()

	plis := bufconn.Listen(1 << 20)
	psrv := grpc.NewServer()
	storagev1.RegisterStorageServiceServer(psrv, newServer())
	go func() { _ = psrv.Serve(plis) }()
	t.Cleanup(psrv.Stop)
	providerConn := bufDial(t, plis)

	core := newGateway(providerConn, "rat-strategy-test", []string{"rat://storage/v1/vend-credentials"})

	glis := bufconn.Listen(1 << 20)
	gsrv := grpc.NewServer()
	corev1.RegisterCapabilityInvokeServiceServer(gsrv, core)
	go func() { _ = gsrv.Serve(glis) }()
	t.Cleanup(gsrv.Stop)

	return &rig{gw: corev1.NewCapabilityInvokeServiceClient(bufDial(t, glis)), core: core, tenant: tenant}
}

func tctx(t *testing.T) context.Context {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	t.Cleanup(cancel)
	return ctx
}

// withCallMeta attaches the rat-callmeta-bin envelope (ADR-007) carrying the
// caller's tenant — which storage reads to scope the creds — and a well-formed
// traceparent (or the gateway rejects the call).
func (r *rig) withCallMeta(ctx context.Context) context.Context {
	rc := &commonv1.RequestContext{
		Trace: &commonv1.TraceContext{
			Traceparent:   "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
			CorrelationId: "corr-golden",
		},
		Identity: &commonv1.Identity{Tenant: r.tenant},
	}
	b, _ := proto.Marshal(rc)
	return metadata.AppendToOutgoingContext(ctx, callMetaHeader, string(b))
}

func (r *rig) vend(ctx context.Context, prefix, mode string) (*storagev1.VendCredentialsResponse, error) {
	payload, err := proto.Marshal(&storagev1.VendCredentialsRequest{Prefix: prefix, Mode: modeEnum(mode)})
	if err != nil {
		return nil, err
	}
	out, err := r.gw.Invoke(r.withCallMeta(ctx), &corev1.InvokeRequest{
		Capability: "rat://storage/v1/vend-credentials", Payload: payload,
	})
	if err != nil {
		return nil, err
	}
	var resp storagev1.VendCredentialsResponse
	if err := proto.Unmarshal(out.GetResult(), &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ---- the tests ----

func TestStorageConformance_GoldenVectors(t *testing.T) {
	v := loadVectors(t)
	r := newRig(t, v.Tenant)
	ctx := tctx(t)
	ttlMs := v.CredentialsTTLs * 1000
	const slackMs = 2000

	for _, s := range v.Lifecycle {
		s := s
		t.Run(s.Step, func(t *testing.T) {
			before := nowMs()
			resp, err := r.vend(ctx, s.Prefix, s.Mode)
			after := nowMs()
			if err != nil {
				t.Fatalf("vend: %v", err)
			}
			if s.Expect.CredentialsPresent && len(resp.GetCredentials()) == 0 {
				t.Fatalf("credentials empty, want present")
			}
			// Decode the conformance scope receipt and assert the C7 binding.
			var got scopeReceipt
			if err := json.Unmarshal(resp.GetCredentials(), &got); err != nil {
				t.Fatalf("decode scope receipt: %v", err)
			}
			if e := s.Expect.Scope; e != nil {
				if got.Tenant != e.Tenant {
					t.Fatalf("scope.tenant = %q, want %q", got.Tenant, e.Tenant)
				}
				if got.Prefix != e.Prefix {
					t.Fatalf("scope.prefix = %q, want %q", got.Prefix, e.Prefix)
				}
				if got.Mode != e.Mode {
					t.Fatalf("scope.mode = %q, want %q", got.Mode, e.Mode)
				}
			}
			// Short-TTL: expiry ~ now + TTL (bounded both ways).
			exp := resp.GetExpiresUnixMs()
			if exp < before+ttlMs-slackMs || exp > after+ttlMs+slackMs {
				t.Fatalf("expires_unix_ms = %d, want within [%d, %d]", exp, before+ttlMs-slackMs, after+ttlMs+slackMs)
			}
			if exp != got.ExpiresUnixMs {
				t.Fatalf("response expires %d != receipt expires %d", exp, got.ExpiresUnixMs)
			}
		})
	}

	if got, want := len(r.core.auditLog()), len(v.Lifecycle); got != want {
		t.Fatalf("audit log = %d entries, want one per mediated call (%d)", got, want)
	}
}

func TestStorageConformance_ErrorVectors(t *testing.T) {
	v := loadVectors(t)
	r := newRig(t, v.Tenant)
	ctx := tctx(t)

	for _, s := range v.Errors {
		s := s
		t.Run(s.Step, func(t *testing.T) {
			_, err := r.vend(ctx, s.Prefix, s.Mode)
			if err == nil {
				t.Fatalf("%s: want error %s, got nil", s.Step, s.Expect.Code)
			}
			if got := status.Code(err); got != wantCode(t, s.Expect.Code) {
				t.Fatalf("%s: status = %s, want %s", s.Step, got, s.Expect.Code)
			}
		})
	}
}

// TestStorage_TenantComesFromMetadataNotRequest is the C7 structural check: the
// vended tenant tracks the rat-callmeta-bin envelope, and there is no request
// field that could override it. Vending under a different caller tenant yields
// creds scoped to THAT tenant — proving the binding is to metadata, not the body.
func TestStorage_TenantComesFromMetadataNotRequest(t *testing.T) {
	r := newRig(t, "globex") // different tenant than the vectors' "acme"
	ctx := tctx(t)
	resp, err := r.vend(ctx, "s3://bucket/x", "READ")
	if err != nil {
		t.Fatalf("vend: %v", err)
	}
	var got scopeReceipt
	if err := json.Unmarshal(resp.GetCredentials(), &got); err != nil {
		t.Fatalf("decode: %v", err)
	}
	if got.Tenant != "globex" {
		t.Fatalf("scope.tenant = %q, want globex (the metadata tenant)", got.Tenant)
	}
}

func wantCode(t *testing.T, name string) codes.Code {
	t.Helper()
	switch name {
	case "INVALID_ARGUMENT":
		return codes.InvalidArgument
	case "PERMISSION_DENIED":
		return codes.PermissionDenied
	case "NOT_FOUND":
		return codes.NotFound
	default:
		t.Fatalf("unmapped expected code %q", name)
		return codes.Unknown
	}
}
